discord/commands/mixins.py

```diff
@@ -0,0 +1,910 @@
+        self.enabled: bool = kwargs.get("enabled", True)
+        self.cooldown_after_parsing: bool = kwargs.get("cooldown_after_parsing", False)
+
+        # checks
+        if checks := getattr(func, "__commands_checks__", []):
+            checks.reverse()
+
+        checks += kwargs.get(
+            "checks", []
+        )  # combine all the checks we find (kwargs or decorator)
+        self.checks: list[Check] = checks
+
+        # cooldowns
+        cooldown = getattr(func, "__commands_cooldown__", kwargs.get("cooldown"))
+
+        if cooldown is None:
+            buckets = CooldownMapping(cooldown, BucketType.default)
+        elif isinstance(cooldown, CooldownMapping):
+            buckets = cooldown
+        else:
+            raise TypeError(
+                "Cooldown must be a an instance of CooldownMapping or None."
+            )
+
+        self._buckets: CooldownMapping = buckets
+
+        # max concurrency
+        self._max_concurrency: MaxConcurrency | None = getattr(
+            func, "__commands_max_concurrency__", kwargs.get("max_concurrency")
+        )
+
+        # hooks
+        self._before_invoke: Hook | None = None
+        if hook := getattr(func, "__before_invoke__", None):
+            self.before_invoke(hook)
+
+        self._after_invoke: Hook | None = None
+        if hook := getattr(func, "__after_invoke__", None):
+            self.after_invoke(hook)
+
+        self.on_error: Error | None
+
+    @property
+    def callback(self) -> CallbackT:
+        return self._callback
+
+    @callback.setter
+    def callback(self, func: CallbackT) -> None:
+        if not asyncio.iscoroutinefunction(func):
+            raise TypeError("Callback must be a coroutine.")
+
+        self._callback = func
+        unwrap = unwrap_function(func)
+        self.module = unwrap.__module__
+
+    @property
+    def cooldown(self):
+        return self._buckets._cooldown
+
+    @property
+    def qualified_name(self) -> str:
+        """:class:`str`: Retrieves the fully qualified command name.
+
+        This is the full parent name with the command name as well.
+        For example, in ``?one two three`` the qualified name would be
+        ``one two three``.
+        """
+        if not self.parent:
+            return self.name
+
+        return f"{self.parent.qualified_name} {self.name}"
+
+    @property
+    def cog_name(self) -> str | None:
+        """Optional[:class:`str`]: The name of the cog this command belongs to, if any."""
+        return type(self.cog).__cog_name__ if self.cog is not None else None
+
+    @property
+    def parents(self) -> list[Invokable]:
+        """List[:class:`Invokable`]: Retrieves the parents of this command.
+
+        If the command has no parents then it returns an empty :class:`list`.
+
+        For example in commands ``?a b c test``, the parents are ``[c, b, a]``.
+        """
+        entries = []
+        command = self
+        while command.parent is not None:  # type: ignore
```

Comment:
Leave a comment why you have `type: ignore` comments here.

Reply:
Looks like command is now solidly typed.